src/MizarData/MizarDataTest.cpp

```diff
@@ -0,0 +1,138 @@
+
+constexpr size_t kTimersNum = 5;
+constexpr std::array<uint64_t, kTimersNum> kStarts{10, 20, 30, 40, 50};
+constexpr std::array<uint64_t, kTimersNum> kEnds{110, 220, 330, 440, 550};
+// These are the timer_info types that are stored in MizarData
+constexpr std::array<TimerInfo::Type, kTimersNum> kTypesToStore{
+    TimerInfo::kNone, TimerInfo::kNone, TimerInfo::kNone, TimerInfo::kApiScope,
+    TimerInfo::kApiScope};
+constexpr std::array<TimerInfo::Type, kTimersNum> kTypesToIgnore{
+    TimerInfo::kGpuActivity, TimerInfo::kApiEvent, TimerInfo::kApiEvent, TimerInfo::kApiScopeAsync,
+    TimerInfo::kApiEvent};
+constexpr uint64_t kFunctionId = 1;
+const std::string kFunctionName = "foo()";
+const std::string kMSName = "ManualScope";
+constexpr uint32_t kTID = 123;
+
+static std::array<TimerInfo, kTimersNum> MakeTimerInfos(
+    const std::array<TimerInfo::Type, kTimersNum>& types) {
+  std::array<TimerInfo, kTimersNum> result;
+  for (size_t i = 0; i < kTimersNum; ++i) {
+    result[i].set_thread_id(kTID);
+    result[i].set_start(kStarts[i]);
+    result[i].set_end(kEnds[i]);
+    TimerInfo::Type type = types[i];
+    result[i].set_type(type);
+    if (type == TimerInfo::kNone) {
+      result[i].set_function_id(kFunctionId);
+    } else {
+      result[i].set_api_scope_name(kMSName);
+    }
+  }
+  return result;
+}
+
+const std::array<TimerInfo, kTimersNum> kTimersToStore = MakeTimerInfos(kTypesToStore);
+const std::array<TimerInfo, kTimersNum> kTimersToIgnore = MakeTimerInfos(kTypesToIgnore);
+
+const orbit_grpc_protos::CaptureStarted kCaptureStarted = [] {
+  orbit_grpc_protos::CaptureStarted result;
+  orbit_grpc_protos::InstrumentedFunction* function =
+      result.mutable_capture_options()->add_instrumented_functions();
+  function->set_function_id(kFunctionId);
+  function->set_function_name(kFunctionName);
+  return result;
+}();
+
+absl::flat_hash_set<orbit_client_data::ScopeType> kStoredScopeTypes = {
+    orbit_client_data::ScopeType::kApiScope,
+    orbit_client_data::ScopeType::kDynamicallyInstrumentedFunction};
+
+static void CallOnCaptureStarted(MizarData& data) {
+  orbit_grpc_protos::CaptureStarted capture_started;
+  capture_started.capture_options().instrumented_functions();
+  data.OnCaptureStarted(kCaptureStarted, "path/to/file", {});
+}
+
+bool operator==(const TimerInfo a, const TimerInfo b) {
+  return a.thread_id() == b.thread_id() && a.start() == b.start() && a.end() == b.end() &&
+         a.function_id() == b.function_id() && a.api_scope_name() == b.api_scope_name() &&
+         a.type() == b.type() && a.thread_id() == b.thread_id();
+}
+
+MATCHER(TimerInfosEq, "") {
+  const TimerInfo& a = std::get<0>(arg);
+  const TimerInfo& b = std::get<1>(arg);
+
+  return a.thread_id() == b.thread_id() && a.start() == b.start() && a.end() == b.end() &&
+         a.function_id() == b.function_id() && a.api_scope_name() == b.api_scope_name() &&
+         a.type() == b.type() && a.thread_id() == b.thread_id();
```

Comment:
I see two ways to improve that:

- Either use `MessageDifferencer::Equals` (which simplifies the code)
- Or express it in terms of a tree of matchers (which means it will give you the exact field that doesn't match when it doesn't match)